### PR TITLE
[gis] new binaryToGeoJSON signature

### PR DESCRIPTION
**Pull request by Pessimistress**
For discussions in https://github.com/visgl/deck.gl/pull/5965#discussion_r667335516

- Rename `binaryToGeoJson` to `binaryToGeoJSON`
- Change `binaryToGeoJson` signature (see documentation)
- Add `binaryToGeometry` export

### CODE REVIEW

**Comment by Pessimistress** [comment]
The gis module is not publicly documented. Do I still need to add this to upgrade guide?

File: modules/gis/src/lib/binary-to-geojson.ts [hunk 1]
```diff
@@ -20,34 +20,36 @@ import type {
...
- * @param type   Input data type: Point, LineString, or Polygon
- * @param format Output format, either geometry or feature
+ * @param options
+ * @param options.type  Input data type: Point, LineString, or Polygon
+ * @param options.featureId  Global feature id. If specified, only a single feature is extracted
  * @return GeoJSON objects
  */
-export function binaryToGeoJson(
+export function binaryToGeoJSON(
   data: BinaryGeometry | BinaryFeatures,
```
Comment (felixpalmer) [inline modules/gis/src/lib/binary-to-geojson.ts:37]: I thought we said we would only support `data: BinaryFeatures` and restrict the return type to `Feature[] | Feature` now that we have `parseGeometry`

Comment (felixpalmer) [reply modules/gis/src/lib/binary-to-geojson.ts:37]: In fact, unless I'm missing something, if we take that line of thinking to its' logical conclusion, perhaps we want to remove the `isHeterogeneousType` legacy check and remove the `type` from `BinaryToGeoJsonOptions` as it just seems to be a roundabout way to invoking `parseGeometry`

**Review by felixpalmer — CHANGES_REQUESTED** [review]

File: modules/geopackage/src/lib/parse-geopackage.ts [hunk 2]
```diff
@@ -15,7 +15,7 @@ import {
   Float32,
   Binary
 } from '@loaders.gl/schema';
-import {binaryToGeoJson, transformGeoJsonCoords} from '@loaders.gl/gis';
+import {binaryToGeoJSON, transformGeoJsonCoords} from '@loaders.gl/gis';
```
Comment (ibgreen) [inline modules/geopackage/src/lib/parse-geopackage.ts:18]: Let's be consistent with GeoJson capitalization. We shouldn't change it in just one place, suggest leaving it lowercase (i.e the capitalizated form of `geojson` in `geojsonToBinary`)

File: modules/gis/src/lib/binary-to-geojson.ts [hunk 3]
```diff
@@ -20,34 +20,36 @@ import type {
...
+  return parseFeatures(data, options?.type);
 }
 
 /**
  * Return a single feature from a binary geometry representation as GeoJSON
  * @param data   geometry data in binary representation
  * @return GeoJSON feature
  */
-export function getSingleFeature(data: BinaryFeatures, globalFeatureId: number) {
+function getSingleFeature(data: BinaryFeatures, globalFeatureId: number) {
```
Comment (ibgreen) [inline modules/gis/src/lib/binary-to-geojson.ts:52]: Be explicit about return type.

File: modules/gis/src/lib/binary-to-geojson.ts [hunk 4]
```diff
@@ -72,7 +74,7 @@ export function getSingleFeature(data: BinaryFeatures, globalFeatureId: number)
     }
   }
 
-  return null;
+  throw new Error(`featureId:${globalFeatureId} not found`);
```
Comment (ibgreen) [inline modules/gis/src/lib/binary-to-geojson.ts:77]: Do we want this function to throw? Also overhead of error message...

In a typescript app, typing will alert the user to the fact that the function can return null...

File: modules/gis/src/lib/binary-to-geojson.ts [hunk 5]
```diff
@@ -90,7 +92,11 @@ function parseFeatures(data, type): Geometry | Feature[] {
 }
 
 /** Parse input binary data and return a valid GeoJSON geometry object */
-function parseGeometry(data: BinaryGeometry, startIndex?: number, endIndex?: number): Geometry {
+export function parseGeometry(
```
Comment (ibgreen) [inline modules/gis/src/lib/binary-to-geojson.ts:95]: Hmm I am getting confused. Aren't we already exporting `binaryToGeometry`? What does this do differently?

Comment (Pessimistress) [reply modules/gis/src/lib/binary-to-geojson.ts:95]: This is the function that is exported as `binaryToGeometry`.

Comment (kylebarron) [reply modules/gis/src/lib/binary-to-geojson.ts:95]: I think this is exported as `binaryToGeometry`:

```js
export {binaryToGeoJSON, parseGeometry as binaryToGeometry} from './lib/binary-to-geojson';
```

Maybe simpler to just name this function `binaryToGeometry`.

File: modules/gis/src/lib/binary-to-geojson.ts (hunk 4 above)
Comment (Pessimistress) [reply modules/gis/src/lib/binary-to-geojson.ts:77]: The only scenario that `binaryToGeoJson` could return null is if the `globalFeatureId` option is specified (single feature mode) and not found. If the return type could be null, then even if the user does not use the single feature mode, they still have to check the returned object for null, which adds unnecessary complexity to the application.

**Review by kylebarron — APPROVED** [review]

File: modules/gis/src/lib/binary-to-geojson.ts (hunk 1 above)
Comment (kylebarron) [reply modules/gis/src/lib/binary-to-geojson.ts:37]: I'm happy to restrict `binaryToGeoJson` to just `BinaryFeatures` and use `binaryToGeometry` for `BinaryGeometry` type data.

File: modules/geopackage/src/lib/parse-geopackage.ts (hunk 2 above)
Comment (Pessimistress) [reply modules/geopackage/src/lib/parse-geopackage.ts:18]: Well it was not consistent with `geojsonToBinary` before. I can rename it `binaryToGeojson` (lower case J) instead.

Comment (ibgreen) [reply modules/geopackage/src/lib/parse-geopackage.ts:18]: Yes, that sounds good. 

Agree that the naming is a bit of a mess. ideally if we make a bigger change we should do it across the board.

File: modules/gis/src/lib/binary-to-geojson.ts (hunk 4 above)
Comment (ibgreen) [reply modules/gis/src/lib/binary-to-geojson.ts:77]: Agreed. (Though I supposed that could have been a possible argument for providing separate functions).

File: modules/gis/src/lib/binary-to-geojson.ts (hunk 5 above)
Comment (ibgreen) [reply modules/gis/src/lib/binary-to-geojson.ts:95]: +1

**Review by ibgreen — APPROVED** [review]

File: modules/geopackage/src/lib/parse-geopackage.ts (hunk 2 above)
Comment (felixpalmer) [reply modules/geopackage/src/lib/parse-geopackage.ts:18]: Renamed to binaryToGeojson

File: modules/gis/src/lib/binary-to-geojson.ts (hunk 5 above)
Comment (felixpalmer) [reply modules/gis/src/lib/binary-to-geojson.ts:95]: Renamed to `binaryToGeometry`

File: modules/gis/src/lib/binary-to-geojson.ts (hunk 1 above)
Comment (felixpalmer) [reply modules/gis/src/lib/binary-to-geojson.ts:37]: I've restricted `binaryToGeoJson` to just `BinaryFeatures` but have left the `isHeterogeneousType` check for back-compatibility

**Review by felixpalmer — APPROVED** [review]

**Pull request merged**